### ISSUE TITLE
Do we need to document the need for the cics profile or support command line options for it ?

### ISSUE THREAD

**Issue by ChrisAtkinson1** [IBM/zowe-cli-cics-deploy-plugin#204]
Here - I point to a non existant cics profile and it complains the host property is not found.

C:\Users\CHRISTOPHERAtkinson\Documents\ZoweStuff\ZOWEDEV\PUSH1\tests\Cicsjson\CICSJSON>zowe cics-deploy push bundle --name CICSJS02 --targetdir /u/atkinc/pushtest --zosmf-profile zosmf2c --cics-deploy-profile deploy --ssh-profile ssh2Ckey  --cics-profile bob
Command Preparation Failed:
Cannot read property 'host' of undefine

When I add --host to the command line I get an unknown keyword error 

Eg

 --host winmvs2c.hursley.ibm.com
Command Error:
Unknown argument: host
Command failed due to improper syntax
Command entered: "zowe cics-deploy push bundle --name CICSJS02 --targetdir /u/atkinc/pushtest --zosmf-profile zosmf2c --zosmfProfile zosmf2c --cics-deploy-profile deploy --cicsDeployProfile deploy --ssh-profile ssh2Ckey --sshProfile ssh2Ckey --cics-profile bob --cicsProfile bob --host winmvs2c.hursley.ibm.com"
Available commands are "bundle".
Use "zowe cics-deploy push --help" to view groups, commands, and options.
Error Details:
Unknown argument: host

**Comment by pcoop** [comment]
Ah, okay, yeah... Imperative doesn't complain about unreadable optional profiles (we can experience similar unpleasantness if the optional cics-deploy profile isn't found either). We could make having a cics-profile mandatory I suppose, that should solve the problem; or we can attempt to detect this type of usage error and issue a better message from within our code.

**Comment by pcoop** [comment]
I've had a quick play and unfortunately we can't issue a better message from our code - our handler never even runs in this scenario so we don't get the chance to find the problem. Our options are either to make a cics-profile mandatory or to seek a solution somewhere in Imperative.

I imagine the the Imperative folks will be unimpressed; they want us to have command line parameters to duplicate each field from the profiles; if we had such parameters then a different layer of validation should detect the problem. At least, I imagine that's how it's supposed to work.

**Comment by markcocker** [comment]
The parameters `host` `port` `user` `password` are duplicated across profiles cics, ssh, and zosmf. If `zowe cics-deploy push bundle` were to have a `--host` option we could override all of these with that value, but `port` certainly wouldn't work so a user is required to have these profiles anyway. We could have unique parameters for port eg. `--port-cics` `--port-ssh` `--port-zosmf`.

cics:
```
--host  | -H (string)
--port  | -P (number)
--user  | -u (string)
--password  | -p (string)
--region-name  (string)
--cics-plex  (string)
--reject-unauthorized  | --ru (boolean)
--protocol  | -o (string)
--response-format-json  | --rfj (boolean)
```
ssh:
```
--host  | -H (string)
--port  | -P (number)
--user  | -u (string)
--password  | --pass | --pw (string)
--privateKey  | --key | --pk (string)
--keyPassphrase  | --passphrase | --kp (string)
--handshakeTimeout  | --timeout | --to (number)
--response-format-json  | --rfj (boolean)
```
zosmf:
```
--host  | -H (string)
--port  | -P (number)
--user  | -u (string)
--password  | --pass | --pw (string)
--reject-unauthorized  | --ru (boolean)
--base-path  | --bp (string)
--response-format-json  | --rfj (boolean)
```
cics-deploy:
```
--cicsplex  | --cp (string)
--scope  | --sc (string)
--csd-group  | --cg | --csdgroup (string)
--res-group  | --rg | --resgroup (string)
--cics-hlq  | --ch | --cicshlq (string)
--cpsm-hlq  | --cph | --cpsmhlq (string)
--target-directory  | --td | --targetdir | --target-dir (string)
--job-card  | --jc | --jobcard (string)
--response-format-json  | --rfj (boolean)
```

**Comment by ChrisAtkinson1** [comment]
There also seems a bit of difference as to what we use to drive the cmci request.

If I have a bad host or port in the cics-profile - we get no output back eg.

ssh2Ckey  --cics-profile "cics2c badport" --ow
PUSH operation completed

But if we have a bad password or bad userid - it appears that these are either not required - or overwritten by the cics-deploy reporting

ssh2Ckey  --cics-profile "cics2c badpwd" --ow
CICS Regions in Scope 'CALMAS1' of CICSplex 'CAPLEX':
CICS Regions in Scope 'CALMAS1' of CICSplex 'CAPLEX':
   Applid: CALMAS1    jobname: CALMAS1    jobid: JOB73879
CICS Regions in Scope 'CALMAS1' of CICSplex 'CAPLEX':
   Applid: CALMAS1    jobname: CALMAS1    jobid: JOB73879

NODEJSAPP resources for Bundle 'CICSJS02' in Scope 'CALMAS1':
CICS Regions in Scope 'CALMAS1' of CICSplex 'CAPLEX':
   Applid: CALMAS1    jobname: CALMAS1    jobid: JOB73879

NODEJSAPP resources for Bundle 'CICSJS02' in Scope 'CALMAS1':
CICS NODEJSAPP resource 'CICSJSON' is in 'ENABLED' state in region 'CALMAS1' with process id '65778'.
  stdout: /u/atkinc/nodelogs/CALMAS1/CICSJSON/CICSJSON/D20190531.T133025.stdout
  stderr: /u/atkinc/nodelogs/CALMAS1/CICSJSON/CICSJSON/D20190531.T133025.stderr
CICS Regions in Scope 'CALMAS1' of CICSplex 'CAPLEX':
   Applid: CALMAS1    jobname: CALMAS1    jobid: JOB73879

NODEJSAPP resources for Bundle 'CICSJS02' in Scope 'CALMAS1':
CICS NODEJSAPP resource 'CICSJSON' is in 'ENABLED' state in region 'CALMAS1' with process id '65778'.
  stdout: /u/atkinc/nodelogs/CALMAS1/CICSJSON/CICSJSON/D20190531.T133025.stdout
  stderr: /u/atkinc/nodelogs/CALMAS1/CICSJSON/CICSJSON/D20190531.T133025.stderr
PUSH operation completed.

**Comment by matthewpwilson** [comment]
That will depend on whether you have authentication setup in your WUI / SMSS

**Comment by ChrisAtkinson1** [comment]
I don't have authentication.  Guess that explains it.

**Comment by pcoop** [comment]
Mark:

The bigger issue of command line overrides gets ugly, but I think the following might just about work.

We'd have to define the external parameters that we want to have on the command line and mark them all as being *optional* in the Imperative files. We then programmatically merge the values from the command line with those from the profiles (we can't rely on Imperative to do this for us because of the clashes); we have to mark everything as optional so that Imperative itself wouldn't complain about missing mandatory values before our Handler even gets control. We would then have a layer of our own validation code behind that to complain if values that we consider to be mandatory are missing.

I think it's doable, but making everything optional even if it's really mandatory might have implications for the command line help, it could be a little ugly.

Imperative itself does have a mechanism to map values from the profiles into differently named command line parameters... but that mapping is in the code that defines the profile and is out of our control. We have to author the merge ourselves. We may have to avoid using any of the clashing parameters in our own externals in order to avoid indeterminacy: if, for example, we had --user has an optional command line parameter and the user didn't set it, we couldn't know which value from which profile had been injected automatically by Imperative for us to use. One value from one of the profiles would win, and we'd essentially override the other two values using it. Whereas, if our command line parameter is --zosuser and it's not set, we could programmatically decide to use the value from the zosMF profile and force that same value to be used for SSH and CMCI work too. That at least would be consistent and something we could document, rather than "we don't know which will win". It's all a bit ugly.


Chris:

As Matthew noted, if security is disabled in CPSM then bad credentials may be tolerated. The bigger problem of how to report failures (e.g. bad host name/port) is something I've struggled with; we may get a simple error message back from the cics plug-in that can be displayed, we might get a binary dump of an HTML page, we might get a large binary representation of an HTTP message... almost anything is possible. In many cases presenting that information is less useful than doing nothing, hence at the moment it does nothing. I'm wondering about adding a basic message of our own, perhaps offering a Zowe command that could be issued to test a CMCI connection, something like that. Any ideas?

**Comment by markcocker** [comment]
I was hoping to avoid proliferating the number of options, but we also want to be compliant. We perhaps need to see advice from the Zowe CLI team.

**Comment by pcoop** [comment]
FWIW, I did raise the issue of profile overrides getting complicated here: https://github.com/zowe/zowe-cli/issues/339 . The reply was _"For your implementation, you could expose everything as CLI options, creating new options where there is overlap. For example, a specific --zosmf-host rather than just --host to differentiate from host in other profiles."_

If we wanted to expose everything to the command line we could perhaps get away with the following additional parameters:

```
--zos-host  | --zh (string) [value from --host in zosMF profile if not set]
--zos-user  | --zu (string) [value from --user in zosMF profile if not set]
--zos-password  | --zp (string) [value from --password in zosMF profile if not set]
--reject-unauthorized  | --ru (boolean) [value from zosMF profile if not set]
--base-path  | --bp (string) [value from zosMF profile if not set]
--handshakeTimeout  | --ht (number) [value from ssh profile if not set]
--protocol  | --pr (string)  [value from the cics profile if not set]
--zosmf-port  | --mp (number) [value from --port in zosMF profile if not set]
--ssh-port  | --sp (number) [value from --port in ssh profile if not set]
--cmci-port | --cics-port | --ep (number) [value from --port in cics profile if not set]

**Comment by matthewpwilson** [comment]
I think it would be clearer to expose all the options from each profile, prefixing each option with the profile name, so:
```
--zosmf-host  | --zh (string) 
--zosmf-user  | --zu (string)
--zosmf-password  | --zpw (string) 
--zosmf-reject-unauthorized  | --zru (boolean) 
--zosmf-base-path  | --zb (string) 
--zosmf-port  | --zp (number)

--ssh-host  | --sh (string) 
--ssh-port  | --sp (number)
--ssh-user | --su (string)  
--ssh-password | --spw (string)
--ssh-handshake-timeout  | --sht (number) 
--ssh-private-key |--spk (string)
--ssh-key-passphrase | --skp (string)

--cics-host | --ch (string)
--cics-port | --cpo (number)
--cics-user | --cu (string)
--cics-password | --cpw (string)
--cics-protocol | --cpr (string)
--cics-reject-unauthorized | --cru (boolean)
```
(cicsplex omitted as we'll always inject that based on the value from the cics-deploy profile)

Command line options should always override values from profiles (this is the convention from Imperative). I think we could also use the values for host, user, password from the zosmf profile or command-line options if the corresponding ssh or cics options are not provided by either profile or CLI. 

For the help I think we can put the options for each profile in its own section as Zowe CLI does with the zosmf options, this avoids cluttering the help for the individual option for each command.

**Comment by pcoop** [comment]
That could work. Password updates will still require three profiles to be updated though, which remains icky. We could validate that if the same --user is shared across multiple profiles then the other credentials must match too, rather than pushing ahead regardless and potentially revoking the id.

We can't have --ch as the short version of both --cics-host and cics-hlq; oh, and perhaps the cics prefix would be better as cmci anyway? Presumably --ssh-handshake-timeout should be --sht in order to match the naming convention, and --ssh-private-key should be --spk.

**Comment by matthewpwilson** [comment]
I prefer `cics-` over `cmci-` so it matches the profile name so it's more obvious the parameters relate to the profile.  Suggest we change the abbreviation for `--cics-hlq` to `--cq` so all the `-host` parameters are consistent.  I've tweaked the others.

**Comment by markcocker** [comment]
@ChrisAtkinson1 From what I can tell this issue has been addressed with PRs #254 #251 #249.

**Comment by ChrisAtkinson1** [comment]
Yes this is now updated to support the cmd line options.
